feat(node): add direct-connection upgrade through relay (dcutr)

aed900 opened the pull request:
Adds DCUtR behaviour to provide hole-punching functionality for upgrading relayed to direct connections. 

Follows on from relay PR: https://github.com/maidsafe/safe_network/pull/666

Yoga07 commented on sn/src/routing/core/msg_handling/agreement.rs:
```diff
@@ -273,28 +275,51 @@ impl Core {
+                // }
             } else {
+
+                let mut section_chain_clone = self.section_chain().await.clone();
```
We should take a copy of the chain before we begin updating. Otherwise, if our prefix gets updated first, the head of the section chain would be our section's and we'd be inserting neighboring section chain head atop ours.

So it's ideal to use old_chain here from L#263.

joshuef replied:
aye sound. im actually pulling in gab's tweak now which needs that too

Yoga07 commented on sn/src/routing/core/msg_handling/anti_entropy.rs:
```diff
@@ -364,6 +383,12 @@ impl Core {
+
+                // assert!(proof_chain.last_key(), );
+
+                trace!(">>>>>  proof chain has our sign key?: {:?}", proof_chain.has_key(&section_auth.public_key_set.public_key()));
```
Do we need these logs and the commented-out codes?

Yoga07 commented on sn/src/routing/core/msg_handling/anti_entropy.rs:
```diff
@@ -43,6 +43,13 @@ impl Core {
+        // FIXME: perhaps we should update our section chain
+        // only upon successful DKG round we participated on??
+        self.section
+            .merge_chain(&signed_section_auth, proof_chain.clone())
```
Would be better to have a check if incoming SAP is our's/

Yoga07 commented on sn/src/routing/core/msg_handling/agreement.rs:
```diff
@@ -271,29 +276,55 @@ impl Core {
-                if let Err(err) = self.section.chain.write().await.insert(
+                let mut section_chain_clone = self.section_chain().await.clone();
+
+                if let Err(err) = section_chain_clone.insert(
```
We can remove this insert part now since it's being covered below.

maqi commented on sn/src/prefix_map/mod.rs:
```diff
@@ -129,17 +129,28 @@ impl NetworkPrefixMap {
-        )
+        );
+
+        for section in self.sections.iter() {
```
```rust
for prefix in self.sections.keys() {
    trace!("Known prefix: {:?}", prefix);
}
```

joshuef replied:
this doesnt work for a dashmap sadly :(

maqi commented on sn/src/routing/core/msg_handling/agreement.rs:
```diff
@@ -271,29 +276,45 @@ impl Core {
+                let network_updated = self.network.verify_with_chain_and_update(
+                    signed_section_auth.clone(),
+                    &proof_chain,
+                    &proof_chain,
```
passing the same chain twice?
bit weird here.
even an empty chain for one of it will be better? if it does being the case.

maqi commented on sn/src/routing/core/msg_handling/anti_entropy.rs:
```diff
@@ -43,6 +43,16 @@ impl Core {
+
+        // FIXME: perhaps we should update our section chain
+        // only upon successful DKG round we participated on??
+        if section_auth.prefix.is_compatible(&our_prefix) {
```
in case we are now in prefix(0) or prefix(1) , and the incoming section_auth is an outdated prefix(),
which this causing any issue?

maqi commented on sn/src/routing/core/msg_handling/anti_entropy.rs:
```diff
@@ -109,11 +113,21 @@ impl Core {
+
+        // If it's out section, update our section details.
+        if self.section.prefix().await == section_auth.prefix {
+            debug!("AE-Retry received from our section, chain was updated so updating our section info");
```
the comment says `chain was updated so updating our section info`
but the following line of code seems just updating the chain?

bochaco commented on sn/src/prefix_map/mod.rs:
```diff
@@ -219,9 +230,9 @@ impl NetworkPrefixMap {
 
         // Check the SAP's key is the last key of the proof chain
-        if proof_chain.last_key() != &signed_section_auth.value.public_key_set.public_key() {
+        if !proof_chain.has_key(&signed_section_auth.value.public_key_set.public_key()) {
```
Do we need to be more laxed? i thought requiring the proof chain to end with the key that belong to the SAP would be easier to reason

joshuef replied:
it shouldn't matter, I think?

It's that or updating the error message to be be more accurate. But merging an old chain should not be possible. 

If it comes from an old key it's still trustworthy, right? just slow?

bochaco replied:
this check is not affected by how old the proof chain or SAP is, it's checking that the proof chain's last key is the key in the SAP, weather they are old or not, it's just about if we accept a proof chain which has more blocks than those needed

bochaco commented on sn/src/routing/core/msg_handling/agreement.rs:
```diff
@@ -258,42 +263,67 @@ impl Core {
             let prefix = section_auth.value.prefix;
             info!("New SAP agreed for {:?}: {:?}", prefix, section_auth);
 
+            let our_prefix = self.section().prefix().await;
```
I think this is a problem, we need to get our prefix before the start of the loop otherwise we will be switching to check other prefix once we update our SAP in a previous iteration of the loop

Yoga07 replied:
We just use this logging below at L280. So I guess that's okay? We are keeping a copy before we update with the new results, so I think this is okay, but yea, taking a copy before would work better.

bochaco replied:
ah, true, then it's not a problem

aed900 committed:
feat(node): add relay server behaviour

Add libp2p's server behaviour to the swarm. Building on PR 101.

aed900 committed:
feat(node): add relay client behaviour and transport

Added relay client behaviour and transport method.

aed900 committed:
feat(node): add direct-connection upgrade through relay (dcutr)

Adds dcutr behaviour for hole-punching functionality to upgrade relayed to direct connections.

joshuef commented:
Closing this for now. We'll keep this in the backpocket, but won't be adding this _as yet_ to avoid overcomplicating things as we stabilise the code.

Thanks for the work here @aed900 🙇
